shell/src/main/java/alluxio/cli/fs/command/DistributedCpCommand.java

```diff
@@ -59,23 +71,85 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
-          conf.get(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT), true,
-          false), 1, mFsContext.getPathConf(dstPath));
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
-      return -1;
-    } finally {
-      thread.interrupt();
+
+    if (PathUtils.hasPrefix(dstPath.toString(), srcPath.toString())) {
+      throw new RuntimeException(ExceptionMessage.MIGRATE_CANNOT_BE_TO_SUBDIRECTORY.getMessage(
+          srcPath, dstPath));
     }
-    System.out.println("Copied " + srcPath + " to " + dstPath);
+
+    mActiveJobs = DEFAULT_ACTIVE_JOBS;
+
+    AlluxioConfiguration conf = mFsContext.getPathConf(dstPath);
+    mWriteType = conf.get(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT);
+
+    distributedCp(srcPath, dstPath);
     return 0;
   }
 
+  private CopyJobAttempt newJob(String srcPath, String dstPath) {
+    CopyJobAttempt jobAttempt = new CopyJobAttempt(mClient,
+        new MigrateConfig(srcPath, dstPath, mWriteType, true, false),
+        new CountingRetry(3));
+
+    jobAttempt.run();
+
+    return jobAttempt;
+  }
+
+  private void distributedCp(AlluxioURI srcPath, AlluxioURI dstPath)
+      throws IOException, AlluxioException {
+    if (mFileSystem.getStatus(srcPath).isFolder()) {
+      createFolders(srcPath, dstPath);
+    }
+    copy(srcPath, dstPath);
+    // Wait remaining jobs to complete.
+    drain();
+  }
+
+  private void createFolders(AlluxioURI srcPath, AlluxioURI dstPath)
+      throws IOException, AlluxioException {
+
+    try {
+      mFileSystem.createDirectory(dstPath);
+      System.out.println("Created directory at " + dstPath.getPath());
+    } catch (FileAlreadyExistsException e) {
+      if (!mFileSystem.getStatus(dstPath).isFolder()) {
+        throw e;
+      }
+    }
+
+    for (URIStatus srcInnerStatus : mFileSystem.listStatus(srcPath)) {
+      if (srcInnerStatus.isFolder()) {
+        String dstInnerPath = computeTargetPath(srcInnerStatus.getPath(),
+            srcPath.getPath(), dstPath.getPath());
+        createFolders(new AlluxioURI(srcInnerStatus.getPath()), new AlluxioURI(dstInnerPath));
+      }
+    }
+  }
+
+  private void copy(AlluxioURI srcPath, AlluxioURI dstPath)
+      throws IOException, AlluxioException {
+
+    for (URIStatus srcInnerStatus : mFileSystem.listStatus(srcPath)) {
+      String dstInnerPath = computeTargetPath(srcInnerStatus.getPath(),
+          srcPath.getPath(), dstPath.getPath());
+      if (srcInnerStatus.isFolder()) {
+        copy(new AlluxioURI(srcInnerStatus.getPath()), new AlluxioURI(dstInnerPath));
+      } else {
+        addJob(srcInnerStatus.getPath(), dstInnerPath);
+      }
+    }
+  }
+
+  private void addJob(String srcPath, String dstPath) {
+    if (mSubmittedJobAttempts.size() >= mActiveJobs) {
+      // Wait one job to complete.
+      waitJob();
+    }
+    mSubmittedJobAttempts.add(newJob(srcPath, dstPath));
+    System.out.println("Copying " + srcPath + " to " + dstPath);
```

Comment:
Did we always print out every file? This could be a huge output, if we are copying a large tree?

Reply:
We didn't but we did for distributedLoad and that has been popular so choosing that instead.